pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/scheduler/RoundRobinScheduler.java

```diff
@@ -52,6 +52,10 @@
         List<Assignment> newAssignments = Lists.newArrayList();
         for (Instance unassignedFunctionInstance : unassignedFunctionInstances) {
             String heartBeatWorkerId = checkHeartBeatFunction(unassignedFunctionInstance);
+            if (heartBeatWorkerId != null && workerIdToAssignment.get(heartBeatWorkerId) == null) {
```

Comment:
I would suggest moving the logic of the heartbeat function outside of the abstraction of the scheduler.  The scheduler is just suppose to assign instances to workers based on some strategy while the heartbeat function scheduling strategy will always be the same.  In the future, if we don't move this logic out and we want to add alternative scheduling strategies, every scheduling strategy will need this logic.  I would suggest moving this logic into the SchedulerManager and just assign it appropriately there.

Reply:
fixed it.

Reply:
I would suggest moving all the logic concerning heartbeat function out of the scheduler.  You can just make create an assignment for the heartbeat function in the scheduler manager.  Ideally we don't want any of this with the heartbeat function within the scheduler

Reply:
I feel any scheduler related responsibility should be part of Scheduler. there will be many things can be shared across multiple implementations. One of the example is "LoadManager" in pulsar where we have multiple implementations and there are many things which are shared in both implementations. and moving responsibilities of such modules at caller might not be good. 
However, moving out heartbeat might not be a big deal and  it will be cleaner as well. let me fix it.